iModelCore/ECPresentation/Source/Shared/Queries/CustomFunctions.cpp

```diff
@@ -447,21 +456,51 @@ struct EvaluateECExpressionScalar : CachingScalarFunction<bmap<ECExpressionScala
             ExpressionContextPtr expressionContext = ECExpressionContextsProvider::GetCalculatedPropertyContext(params);
 
             ECValue value;
-            Utf8String expressionResult;
             ECExpressionsCache noCache;
-            if (ECExpressionsHelper(noCache).EvaluateECExpression(value, expression, *expressionContext) && value.IsPrimitive() && value.ConvertPrimitiveToString(expressionResult))
+            if (!ECExpressionsHelper(noCache).EvaluateECExpression(value, expression, *expressionContext) || !value.IsPrimitive() || !value.ConvertToPrimitiveType(requestedTypePrimitive))
+                {  
+                ctx.SetResultError(Utf8PrintfString("Calculated property evaluated to a type that is not Primitive").c_str());
+                return;
+                }
+
+            if (nullptr != GetContext().GetUsedClassesListener())
                 {
-                if (nullptr != GetContext().GetUsedClassesListener())
-                    {
-                    UsedClassesHelper::NotifyListenerWithUsedClasses(*GetContext().GetUsedClassesListener(),
-                        GetContext().GetSchemaHelper(), expression);
-                    }
+                UsedClassesHelper::NotifyListenerWithUsedClasses(*GetContext().GetUsedClassesListener(),
+                    GetContext().GetSchemaHelper(), expression);
                 }
 
-            iter = GetCache().Insert(key, std::make_shared<Utf8String>(expressionResult)).first;
+            iter = GetCache().Insert(key, std::make_shared<ECValue>(value)).first;
             }
 
-        ctx.SetResultText(iter->second->c_str(), (int)iter->second->size(), BeSQLite::DbFunction::Context::CopyData::No);
+        double expressionResultJulianDay;
+        Utf8String expressionResultStr;
+
+        switch (requestedTypePrimitive)
+            {
+            case (PRIMITIVETYPE_String):
+                iter->second->ConvertPrimitiveToString(expressionResultStr);
+                ctx.SetResultText(expressionResultStr.c_str(), expressionResultStr.size(), BeSQLite::DbFunction::Context::CopyData::Yes);
+                break;
+            case (PRIMITIVETYPE_Integer):
+                ctx.SetResultInt(iter->second->GetInteger());

```

Comment:
what if ECValue stores a string?

Reply:
Then it would not reach this point `ConvertToPrimitiveType` function tells us if it can or cannot convert the evaluated ecvalue to the specified type.